ussologin/login.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package ussologin defines functionality used for allowing clients to authenticate
+// using USSO OAuth.
+package ussologin
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/juju/httprequest"
+	"github.com/juju/usso"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/environschema.v1"
+	"gopkg.in/juju/environschema.v1/form"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+)
+
+type tokenGetter interface {
+	GetTokenWithOTP(username, password, otp, tokenName string) (*usso.SSOData, error)
+}
+
+// This is defined here to allow it to be stubbed out in tests
+var server tokenGetter = usso.ProductionUbuntuSSOServer
+
+// Login completes the login information using the provided filler
+// and attempts to obtain a USSO token using this information.
+// If the store is non-nil it is used to store this token.
+func Login(filler form.Filler, store TokenStore) (*usso.SSOData, error) {
+	login, err := filler.Fill(loginForm)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot read login parameters")
+	}
+	tok, err := server.GetTokenWithOTP(
+		login["username"].(string),
+		login["password"].(string),
+		login["otp"].(string),
+		"charm",
+	)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot get token")
+	}
+	if err := store.Put(tok); err != nil {
+		return nil, errgo.Notef(err, "cannot save token")
+	}
+	return tok, nil
+}
+
+var loginForm = form.Form{
+	Fields: environschema.Fields{
+		"username": environschema.Attr{
```

Comment:
I think the keys in this map will be used as the prompts in the filler by default. You may want to give them more meaningful names. however you can try it and see and always just update if they're a little terse.